File: Comportamentales Python/observer.py
# Observer Pattern
from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self, subject):
        pass

class Subject(ABC):
    def __init__(self):
        self._observers = []
    
    def attach(self, observer: Observer):
        self._observers.append(observer)
    
    def detach(self, observer: Observer):
        self._observers.remove(observer)
    
    def notify(self):
        for observer in self._observers:
            observer.update(self)

class ConcreteSubject(Subject):
    _state = None
    
    @property
    def state(self):
        return self._state
    
    @state.setter
    def state(self, value):
        self._state = value
        self.notify()

class ConcreteObserverA(Observer):
    def update(self, subject: Subject):
        if isinstance(subject, ConcreteSubject) and subject.state < 3:
            print("ConcreteObserverA: Reacted to the event")

class ConcreteObserverB(Observer):
    def update(self, subject: Subject):
        if isinstance(subject, ConcreteSubject) and (subject.state == 0 or subject.state >= 2):
            print("ConcreteObserverB: Reacted to the event")

# Usage
if __name__ == "__main__":
    subject = ConcreteSubject()
    
    observer_a = ConcreteObserverA()
    subject.attach(observer_a)
    
    observer_b = ConcreteObserverB()
    subject.attach(observer_b)
    
    subject.state = 1
    subject.state = 2

File: Comportamentales Java/Observer.java
// Observer Pattern
import java.util.ArrayList;
import java.util.List;

interface Observer {
    void update(Subject subject);
}

abstract class Subject {
    private List<Observer> observers = new ArrayList<>();
    
    public void attach(Observer observer) {
        observers.add(observer);
    }
    
    public void detach(Observer observer) {
        observers.remove(observer);
    }
    
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(this);
        }
    }
}

class ConcreteSubject extends Subject {
    private int state;
    
    public int getState() {
        return state;
    }
    
    public void setState(int state) {
        this.state = state;
        notifyObservers();
    }
}

class ConcreteObserverA implements Observer {
    @Override
    public void update(Subject subject) {
        if (subject instanceof ConcreteSubject) {
            ConcreteSubject concreteSubject = (ConcreteSubject) subject;
            if (concreteSubject.getState() < 3) {
                System.out.println("ConcreteObserverA: Reacted to the event");
            }
        }
    }
}

class ConcreteObserverB implements Observer {
    @Override
    public void update(Subject subject) {
        if (subject instanceof ConcreteSubject) {
            ConcreteSubject concreteSubject = (ConcreteSubject) subject;
            if (concreteSubject.getState() == 0 || concreteSubject.getState() >= 2) {
                System.out.println("ConcreteObserverB: Reacted to the event");
            }
        }
    }
}

// Usage
class ObserverDemo {
    public static void main(String[] args) {
        ConcreteSubject subject = new ConcreteSubject();
        
        Observer observerA = new ConcreteObserverA();
        subject.attach(observerA);
        
        Observer observerB = new ConcreteObserverB();
        subject.attach(observerB);
        
        subject.setState(1);
        subject.setState(2);
    }
}
